core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,298 @@
+    this.heartbeatProperties = new ConcurrentHashMap<>();
+    this.isEnabled = true;
+    this.heartbeatsSent = new AtomicLong(0);
+  }
+
+  @Override
+  public String getInstrumentationKey() {
+    return this.telemetryClient.getContext().getInstrumentationKey();
+  }
+
+  @Override
+  public void setInstrumentationKey(String key) {
+    if (this.telemetryClient != null && key != null) {
+      this.telemetryClient.getContext().setInstrumentationKey(key);
+    }
+  }
+
+  @Override
+  public void initialize(TelemetryConfiguration configuration) {
+    if (this.telemetryClient == null) {
+      this.telemetryClient = new TelemetryClient(configuration);
+    }
+
+    //Submit task to set properties to dictionary using separate thread. we do not wait for the
+    //results to come out as some I/O bound properties may take time.
+    executorService.submit(HeartbeatDefaultPayload.populateDefaultPayload(getExcludedHeartBeatProperties(),
+        getExcludedHeartBeatPropertyProviders(), this));
+
+
+    if (isEnabled) {
+      scheduledExecutorService.scheduleAtFixedRate(heartBeatPulse(), interval, interval, TimeUnit.SECONDS);
+    }
+
+  }
+
+  @Override
+  public boolean addHeartBeatProperty(String propertyName, String propertyValue,
+      boolean isHealthy) {
+
+    boolean isAdded= false;
+    if (!StringUtils.isEmpty(propertyName)) {
+      try {
+        if (!heartbeatProperties.containsKey(propertyName)) {
+             HeartBeatPropertyPayload payload = new HeartBeatPropertyPayload();
+             payload.setHealthy(isHealthy);
+             payload.setPayloadValue(propertyValue);
+             heartbeatProperties.put(propertyName, payload);
+             isAdded = true;
+             InternalLogger.INSTANCE.trace("added heartbeat property");
+        }
+        else {
+          throw new Exception("heartbeat property cannot be added twice");
+        }
+      } catch (Exception e) {
+        InternalLogger.INSTANCE.warn("Failed to add the property %s value %s, stack trace is : %s," ,
+            propertyName, propertyValue, ExceptionUtils.getStackTrace(e));
+      }
+    }
+    else {
+      InternalLogger.INSTANCE.warn("cannot add property without property name");
+    }
+    return isAdded;
+  }
+
+  @Override
+  public boolean setHeartBeatProperty(String propertyName, String propertyValue,
+      boolean isHealthy) {
+
+    boolean setResult = false;
+    if (!StringUtils.isEmpty(propertyName)) {
+      try {
+        if (heartbeatProperties.containsKey(propertyName) && !HeartbeatDefaultPayload.isDefaultKeyWord(propertyName)) {
+          HeartBeatPropertyPayload payload = new HeartBeatPropertyPayload();
+          payload.setHealthy(isHealthy);
+          payload.setPayloadValue(propertyValue);
+          heartbeatProperties.put(propertyName, payload);
+          setResult = true;
+        }
+        else {
+          throw new Exception("heartbeat property cannot be set without adding it first, or heartbeat"
```

Comment:
Given the chance to do this a second time, I would allow Add *or* Set to actually add new fields. Perhaps issue a logger warning if a user is Adding new values via Set, if there is any real danger of doing this.

Originally, I was concerned that allowing adds to occur at any time might encourage the abuse of the Heartbeat for what is more appropriate to do with TrackEvent calls. And with the heartbeats properties being around for the entire duration of a program's runtime, I was slightly concerned with memory fragmentation should any abuse occur. However, [I doubt there is much of a real problem there](https://stackoverflow.com/questions/5240860/what-causes-memory-fragmentation-in-net).

Reply:
@d3r3kk  Actually, I would like to keep the convention of add by default for adding and set by default for modifying. So keeping this in mind logically add should come before modifying (aka set in this case). I agree this is a bit opinionated and restricting the users but setting the convention I think will prove to be better in long run. Lot's of tunable knobs sometimes lead to spaghetti :)

Reply:
@dhaval24, I'm inclined to agree with @d3r3kk. See the methods defined in Java's Map class. There could be places in the code where you won't know if a key has been added or not. I'm not a fan of the _if containsKey **set(x=y)** else **add(x=y)**_ pattern when you can have a simple _**set(x=y)**_.

Just as in Java's Map, putIfAbsent can be nice, if there's a precedent.

Reply:
@littleaj this is addressed